lib/CountdownEvent.php

```diff
@@ -0,0 +1,59 @@
+<?php
+
+namespace Amp;
+
+use InvalidArgumentException;
+use RuntimeException;
+
+/**
+ * A countdown event is a promise that is resolved when it was signaled n times
+ *
+ * **Example**
+ *
+ * ```php
+ * $countdownEvent = new CountdownEvent(2);
+ * $countdownEvent->signal();
+ * $countdownEvent->signal(); // promise is now resolved
+ * ```
+ */
+final class CountdownEvent implements Promise
+{
+    /** @var int */
+    private $counter;
+    /** @var Deferred */
+    private $deferred;
+
+    public function __construct(int $counter)
+    {
+        if ($counter < 1) {
+            throw new InvalidArgumentException('Counter must be positive');
+        }
+
+        $this->counter = $counter;
+        $this->deferred = new Deferred();
+    }
+
+    /**
+     * @return void
+     */
+    public function signal()
+    {
+        if (0 === $this->counter) {
+            throw new RuntimeException('CountdownEvent already resolved');
+        }
+
+        --$this->counter;
+
+        if (0 === $this->counter) {
+            $this->deferred->resolve(true);
+        }
+    }
```

Comment:
We usually separate the objects that resolve a promise and the object that is passed to the API user to subscribe to.

Reply:
So what would you suggest here instead?

Reply:
Having a `promise()` method like `Deferred` that returns the promise to the consumer. Or maybe it's fine here, since these objects aren't supposed to be passed across boundaries? Dunno.